### ISSUE TITLE
npm v3 dep resolution doc

### ISSUE THREAD

**Issue by RichardLitt** [npm/docs#513]
I have some questions!

- How do you decide which version to install at top level? Is this just alphabetical, based on whether or not A has dep B before C is installed with dep-b Bv2 inside? 
- How do I see what the equivalent of what I used to see inside of node_modules: modules used or previously installed as dependencies of my module, but not as sub-deps? Say I used to require A and B inside of my index.js, and I had previously also required C without saving it, so that I used to be able to look in node_modules to see A, B, and C. Now I see a ton of more dependencies, most of which aren't used in my index.js file. How do I see what I used to see?
- Why the change?

**Comment by maxogden** [comment]
+1 to 'Why the change?'

**Comment by ashleygwilliams** [comment]
working on an answer for you but in the meantime note npm packages `mod-a`, `mod-b`, `mod-c`, and `mod-d`, they follow the examples in the docs, so you can play around with them.

`mod-d` is not yet in the docs but is similar to C in that it depends on Bv2.

i'm working on a sandbox to demo these: https://github.com/ashleygwilliams/npm-sandbox

**Comment by ashleygwilliams** [comment]
generally, the answers are:

1. install order. this is *primarily* alphabetical since `npm install` orders packages alphabetically, but as you might be able to imagine there are simple ways to make this non-deterministic. a `package.json` in the absence of a local `node_modules` folder will always produce the same results, alphabetical(which is the same as install order).

2. from my knowledge... you can't anymore. but i'm not sure so i'm investigating.

3. this is also going to take a second. some preliminary thoughts are that it does reduce tree depth and some duplication.

**Comment by ashleygwilliams** [comment]
@RichardLitt ^ updated comment about seeing just primary deps

**Comment by RichardLitt** [comment]
Thanks for taking the time to answer! Number 2, in particular, was confusing to me, good to know how to do that. 

> as you might be able to imagine there are simple ways to make this non-deterministic

Can you help me out with this? What would you suggest, here? It would seem to me that they ought to be installed at a level via versioning, instead of the random alphabetical install order of the 0-depth packages (alphabetical = random, as far as semantic ordering goes).

**Comment by ashleygwilliams** [comment]
so-- i am planning on adding a page specifically about this #518 

if you have specific questions, please file them there so i can be sure to answer :)

**Comment by RichardLitt** [comment]
Cool. Will do. Nothing more at the moment. Let us know when the sandbox is done, I'd be keen to test it out. :+1:

**Comment by ashleygwilliams** [comment]
so @RichardLitt i just shipped https://docs.npmjs.com/how-npm-works/npm3-dupe, would love to know if that helps! also the sandbox is here, though not fully updated yet: https://github.com/ashleygwilliams/npm-sandbox

i'm trying to set it up so that there is a directory for every example in the docs and that the doc pages are copied over so the explanation is local.

**Comment by ashleygwilliams** [comment]
you also might checkout https://docs.npmjs.com/how-npm-works/npm3-nondet which specifically addresses non-deterministic behavior

**Comment by RichardLitt** [comment]
Cool; thanks! I understand how the current process works. 

If you don't mind, I have a couple of other questions:
- Why is alphabetical order enforced if a primary concern is deduplication? Wouldn't it be smarter to have the most used versions of a dep installed at top level, rather than the arbitrary version that the installer first runs across?
- What if I want to edit a dep itself that is used be several modules; can I copy the dep into the folder for the module I want it to access, and then will that dep then be used by that module? Is there a special de-link command I need to do this?

**Comment by ashleygwilliams** [comment]
1. i dont have an answer and generally agree that your suggestion would produce a more desirable result. i can talk to the team about it.
2. if i am understanding you correctly... (which i might not be!) i would do this:
- fork the module (let's call it module A) that you want the edited dep on
- fork the dep (let's call it module B) you want to edit 
- in your app's `package.json` point module A to your fork of module A
- in module A's `package.json` point to your fork of module B
- edit module B

this would result in:

-  all the app's deps that depended on the non-edited version of module B still dep on a non-edited version of module B.
- module A depends on an edited version of module B

the `node_modules` directory will, again, depend on install order. it is my understanding that if module A was installed first then the edited version of module B would be top level and the other version of module B that the other deps required would be nested below each.

let me know if that helps!

**Comment by RichardLitt** [comment]
1. Cool. Looking forward to hearing about that. I think alphabetization is a stupid way to do anything not-human-readable, so it would be great if we can move away from it. 

2 . Yes, I understand that, although it makes editing deps a lot harder than it used to be, and means that things like `npm shrinkwrap` aren't going to be as straight forward, I think. For instance, editing a dep at say, 5th or 6th level down now means I need to fork... a few projects? Whereas before it just mean editing the files locally. I may have misunderstood. I can write up a more succinct question and explanation of my view, if you'd like.

**Comment by ashleygwilliams** [comment]
2. yes please do! probably open a new issue for it?

**Comment by ashleygwilliams** [comment]
closing. would very happily welcome a new issue on that second subject.

**Comment by RichardLitt** [comment]
@ashleygwilliams Will do! Sorry, been busy the past few days.

**Comment by ashleygwilliams** [comment]
lol no worries, monday is just my triage day :) @RichardLitt
